### PR TITLE
Fix another clashing_extern_declarations false positive.

### LINKED ISSUE #75739: Additional incorrect clashing_extern_decl warning for `#[repr(transparent)]` non-zero types
It seems that there is a case still not being fixed, when `#[repr(transparent)]` is used with non-zero types.
```rust
mod a {
    use std::num::NonZeroUsize;
    extern "C" {
        fn a() -> NonZeroUsize;
    }
}

mod b {
    #[repr(transparent)]
    struct X(NonZeroUsize);
    use std::num::NonZeroUsize;
    extern "C" {
        fn a() -> X;
    }
}
```

_Originally posted by @nbdd0121 in https://github.com/rust-lang/rust/issues/73735#issuecomment-677152928_

### PR DESCRIPTION
**Pull request by jumbatm**
Fixes #75739.

Fix another clashing_extern_declarations false positive, this time for transparent newtype with a non-zero member.

r? @lcnr

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
I'm going to nominate this for beta backport, similarly to https://github.com/rust-lang/rust/pull/73990. However, note:

 * 1.46 received a beta backport simply allowing this lint by default
 * 1.47 (current nightly, soon to be promoted to beta) has the lint on by default

I suspect that it might make sense to continue allowing the lint by default for another cycle, though this patch does seem more minimal than the previous PR which we declined a direct backport of.

We could also say this is "just a bug" and that it is fine for the false-positive here; I am not clear how likely it is in real-world code to hit this case.

File: src/librustc_lint/builtin.rs [hunk 1]
```diff
@@ -2162,6 +2162,38 @@ impl ClashingExternDeclarations {
...
+                loop {
+                    if let ty::Adt(def, substs) = ty.kind {
+                        let is_transparent = def.subst(tcx, substs).repr.transparent();
+                        let is_enum = def.is_enum();
+                        let is_non_null = crate::types::guaranteed_nonnull_optimization(tcx, &def);
+                        debug!(
+                            "non_transparent_ty({:?}) -- type is transparent? {}, type is enum? {}, type is non-null? {}",
+                            ty, is_transparent, is_enum, is_non_null
+                        );
+                        if is_transparent && !is_enum && !is_non_null {
```
Comment (nagisa) [inline src/librustc_lint/builtin.rs:2180]: enums with a single variant can be valid `#[repr(transparent)]` types I believe.

Comment (jumbatm) [reply src/librustc_lint/builtin.rs]: Ah, true -- thanks!

File: src/librustc_lint/builtin.rs [hunk 2]
```diff
@@ -2162,6 +2163,39 @@ impl ClashingExternDeclarations {
...
+                            "non_transparent_ty({:?}) -- type is transparent? {}, type is non-null? {}",
+                            ty, is_transparent, is_non_null
+                        );
+                        if is_transparent && !is_non_null {
+                            debug_assert!(def.variants.len() == 1);
+                            let v = &def.variants[VariantIdx::new(0)];
+                            assert!(
+                                v.fields.len() > 0,
+                                "single-variant transparent structure with zero-sized field"
+                            );
```
Comment (lcnr) [inline src/librustc_lint/builtin.rs:2186]: This assert is not necessary imo, considering that `transparent_newtype_field` would return `None` in that case

File: src/test/ui/lint/clashing-extern-fn.rs [hunk 3]
```diff
@@ -233,7 +236,7 @@ mod missing_return_type {
             // the correct one -- if this one is the correct one, then calling the usize-returning
             // version would allow reads into uninitialised memory.
             fn missing_return_type();
-            //~^ WARN `missing_return_type` redeclared with a different signature
+        //~^ WARN `missing_return_type` redeclared with a different signature
```
Comment (lcnr) [inline src/test/ui/lint/clashing-extern-fn.rs:239]: there are a bunch of somewhat weird identation changes here, were these intentional?

File: src/librustc_lint/types.rs [hunk 4]
```diff
@@ -527,22 +527,26 @@ enum FfiResult<'tcx> {
     FfiUnsafe { ty: Ty<'tcx>, reason: String, help: Option<String> },
 }
 
+crate fn guaranteed_nonnull_optimization<'tcx>(tcx: TyCtxt<'tcx>, def: &ty::AdtDef) -> bool {
```
Comment (lcnr) [inline src/librustc_lint/types.rs:530]:
```suggestion
crate fn nonnull_optimization_guaranteed<'tcx>(tcx: TyCtxt<'tcx>, def: &ty::AdtDef) -> bool {
```
nit: use the same name as the attribute

**Review by lcnr — APPROVED** [review]
This change looks good to me :+1: 

I kind of feel like this lint is getting somewhat unwieldy though but don't really know how to clean this up myself so :shrug: 

A few minor nits, otherwise r=me

File: src/test/ui/lint/clashing-extern-fn.rs (hunk 3 above)
Comment (jumbatm) [reply src/test/ui/lint/clashing-extern-fn.rs:239]: Oops, no, these were from rustfmt

**Comment by jumbatm** [comment]
Should be good to go once green.

**Comment by lcnr** [comment]
thanks

@bors r+ rollup

**Comment by spastorino** [comment]
[discussed](https://zulip-archive.rust-lang.org/238009tcompilermeetings/49871weeklymeeting2020082754818.html#208222837) in T-compiler meeting, declined to backport.

@rustbot modify labels: -beta-nominated

**Pull request merged**